addons/web/static/tests/legacy/fields/basic_fields_tests.js

```diff
@@ -1321,6 +1321,32 @@ QUnit.module('basic_fields', {
         form.destroy();
     });
 
+    QUnit.test('char widget isValid method works with patterns', async function (assert) {
+        assert.expect(1);
+
+        this.data.partner.fields.foo.pattern = 'azerty';
+        var form = await createView({
+            View: FormView,
+            model: 'partner',
+            data: this.data,
+            arch:'<form string="Partners">' +
+                        '<field name="foo"/>' +
+                '</form>',
+            res_id: 1,
+        });
+
+        await testUtils.form.clickEdit(form);
+        await testUtils.fields.editInput(form.$('input[type="text"].o_field_widget'), 'this is wrong');
+        await testUtils.form.clickSave(form);
+
+        // Save should fail, editing to something right this time
```

Comment:
I am right saying you don't *assert* that the edition failed? Shouldn't be a step like `assert.strictEqual(form.$('.o_field_widget').text(), '');`?

Reply:
You shouldn't be able to edit if the save button worked

Reply:
Isn't it a better way to assert the value is wrong? Like a specific css class on the input?

Reply:
you can indeed assert that the field input contains className o_invalid or something. Or at least assert that the view is still in edition, to make it explicit